test/widgets/test_clock.py

```diff
@@ -0,0 +1,330 @@
+
+    # ...but dateutil must be
+    monkeypatch.setitem(sys.modules, "dateutil", True)
+
+    # Override datetime
+    monkeypatch.setattr("libqtile.widget.clock.datetime", MockDatetime)
+
+    # Set up references to pytz and dateutil so we know these aren't being used
+    # If they're called, the widget would try to run None(self.timezone) which
+    # would raise an exception
+    clock.pytz = None
+    clock.dateutil = FakeDateutilTZ
+
+    # Pytz timezone must be a string
+    clk5 = clock.Clock(timezone="1")
+
+    fakebar = Bar([clk5], 24)
+    fakebar.window = FakeWindow()
+    fakebar.width = 10
+    fakebar.height = 10
+    fakebar.draw = no_op
+    clk5._configure(fake_qtile, fakebar)
+    text = clk5.poll()
+
+    # Default time is 10:20 and we add 1 hour for the timezone plus and extra
+    # 1 for the pytz function
+    assert text == "13:20"
+
+
+def test_clock_tick(manager_nospawn, minimal_conf_noscreen, monkeypatch):
+    ''' Test clock ticks '''
+
+    class FakeDateutilTZ:
+        class TZ:
+            @classmethod
+            def gettz(cls, val):
+                return int(val) + 2
+
+        tz = TZ
+
+    class TickingDateTime(datetime.datetime):
+        offset = 0
+
+        @classmethod
+        def now(cls, *args, **kwargs):
+            return cls(2021, 1, 1, 10, 20, 30)
+
+        # This will return 10:20 on first call and 10:21 on all
+        # subsequent calls
+        def astimezone(self, tzone=None):
+            extra = datetime.timedelta(minutes=TickingDateTime.offset)
+            if TickingDateTime.offset < 1:
+                TickingDateTime.offset += 1
+
+            if tzone is None:
+                return self + extra
+            else:
+                return self + datetime.timedelta(hours=tzone) + extra
+
+    # Stop system importing these modules in case they exist on environment
+    monkeypatch.setitem(sys.modules, "pytz", None)
+    monkeypatch.setitem(sys.modules, "dateutil", None)
+    monkeypatch.setitem(sys.modules, "dateutil.tz", None)
+
+    # Reload module to force ImportErrors
+    reload(clock)
+
+    # pytz must not be in the sys.modules dict...
+    monkeypatch.delitem(sys.modules, "pytz")
+
+    # ...but dateutil must be
+    monkeypatch.setitem(sys.modules, "dateutil", True)
+
+    # Override datetime
+    monkeypatch.setattr("libqtile.widget.clock.datetime", TickingDateTime)
+
+    # Set up references to pytz and dateutil so we know these aren't being used
+    # If they're called, the widget would try to run None(self.timezone) which
+    # would raise an exception
+    clock.pytz = None
+    clock.dateutil = FakeDateutilTZ
+
+    # set a short update string so as to tick quickly
+    clk6 = clock.Clock(update_interval=0.01)
+
+    config = minimal_conf_noscreen
+    config.screens = [
+        libqtile.config.Screen(
+            top=libqtile.bar.Bar([clk6], 10)
+        )
+    ]
+
+    manager_nospawn.start(config)
+
+    topbar = manager_nospawn.c.bar["top"]
+    sleep(0.1)
```

Comment:
Let's not fix problems with sleeps. It's just a recipe for unreliable CI in the future.

I see this and one other sleep in the code; I wonder if we can do some synchronization via the IPC interface somehow? Like, having it do callbacks when particular hooks are fired or whatever, so you could fire up IPC and wait for the startup hook to be called.

More magically if we could inspect the qtile object from the test process, that would be handy. I fear that will bring lots of demons, so the above may be easier.

Reply:
Agreed. I was uncomfortable with the `sleep` call too. I actually raised this concern in my vol3 PR (https://github.com/qtile/qtile/pull/2414#discussion_r618102262) and have implemented the fix suggested there.

I can use:
``` python
manager_nospawn.c.widget["clock"].eval("self.tick()")
```

I'm also not a big fan of `eval`!

Your last suggestion would be the nicest but I don't know enough about IPC here. I had hoped a simple:
``` python
def cmd_get_object(self):
    return self
```
would work but it throws out errors and, if it were that easy, it would probably be implemented already,

Reply:
I've fixed this in the `test_check_updates.py` test too. A couple more things to update and then I'll push changes.

I note some other non-widget tests use `sleep` too. I'll add to my to do list and see if we can change those too.

Reply:
> I note some other non-widget tests use sleep too. I'll add to my to do list and see if we can change those too.

Awesome, thanks!

> would work but it throws out errors and, if it were that easy, it would probably be implemented already,

Yeah, the problem is that you need to share the memory across processes, which is not so simple. The qtile process is where the cmd_* stuff gets run, then it gets pickled and sent over a socket to the client process (which is what the test code is).

> I'm also not a big fan of eval!

eval() bothers me less than sleep() since it won't break when the host is under load, but doing some kind of real synchronization would be best. eval() is fine for now, though.